osu.Game.Rulesets.Tau/UI/OrderedHitPolicy.cs

```diff
@@ -0,0 +1,106 @@
+        public bool IsHittable(DrawableHitObject hitObject, double time)
+        {
+            DrawableHitObject blockingObject = null;
+
+            foreach (var obj in enumerateHitObjectsUpTo(hitObject.HitObject.StartTime))
+            {
+                if (hitObjectCanBlockFutureHits(obj))
+                    blockingObject = obj;
+            }
+
+            // If there is no previous hitobject, allow the hit.
+            if (blockingObject == null)
+                return true;
+
+            // A hit is allowed if:
+            // 1. The last blocking hitobject has been judged.
+            // 2. The current time is after the last hitobject's start time.
+            // Hits at exactly the same time as the blocking hitobject are allowed for maps that contain simultaneous hitobjects (e.g. /b/372245).
+            return blockingObject.Judged || time >= blockingObject.HitObject.StartTime;
+        }
+
+        /// <summary>
+        /// Handles a <see cref="HitObject"/> being hit to potentially miss all earlier <see cref="HitObject"/>s.
+        /// </summary>
+        /// <param name="hitObject">The <see cref="HitObject"/> that was hit.</param>
+        public void HandleHit(DrawableHitObject hitObject)
+        {
+            // Hitobjects which themselves don't block future hitobjects don't cause misses (e.g. slider ticks, spinners).
+            if (!hitObjectCanBlockFutureHits(hitObject))
+                return;
+
+            if (!IsHittable(hitObject, hitObject.HitObject.StartTime + hitObject.Result.TimeOffset))
+                throw new InvalidOperationException($"A {hitObject} was hit before it became hittable!");
+
+            foreach (var obj in enumerateHitObjectsUpTo(hitObject.HitObject.StartTime))
+            {
+                if (obj.Judged)
+                    continue;
+
+                if (hitObjectCanBlockFutureHits(obj))
+                    ((DrawableTauHitObject)obj).MissForcefully();
+            }
+        }
+
+        /// <summary>
+        /// Whether a <see cref="HitObject"/> blocks hits on future <see cref="HitObject"/>s until its start time is reached.
+        /// </summary>
+        /// <param name="hitObject">The <see cref="HitObject"/> to test.</param>
+        private static bool hitObjectCanBlockFutureHits(DrawableHitObject hitObject)
+            => hitObject is DrawableBeat || hitObject is DrawableHardBeat;
+
+        private IEnumerable<DrawableHitObject> enumerateHitObjectsUpTo(double targetTime)
+        {
+            foreach (var obj in hitObjectContainer.AliveObjects)
+            {
+                if (obj.HitObject.StartTime >= targetTime)
+                    yield break;
+
+                yield return obj;
+
+                foreach (var nestedObj in obj.NestedHitObjects)
+                {
+                    if (nestedObj.HitObject.StartTime >= targetTime)
+                        break;
+
+                    yield return nestedObj;
+                }
```

Comment:
i'm guessing this is just safety implementation for sliders?

Reply:
Not really. Given that sliders are still a single object.
And Sliders won't really be affected by notelock at all, and they won't be involved at all.

Reply:
I haven't really changed much in here to tailor it for Tau, since I just wanted to get a working notelock. And keeping this block will save us the hassle in the future, should nested objects are used.